perf: Use non-lazy AST traversal for filters when possible

tarleb opened the pull request:
- Add `jog` Lua module
- Use jog instead of walk
- Force use of pandoc walk for some filters
- Add mechanism to check `jog` results

Welcome to the quarto GitHub repo!

We are always happy to hear feedback from our users.

To file a _pull request_, please follow these instructions carefully: <https://yihui.org/issue/#bug-reports>

If you're a collaborator from outside `quarto-dev` making changes larger than a typo, please make sure you have filed an [individual](https://posit.co/wp-content/uploads/2023/04/2023-03-13_TC_Indiv_contrib_agreement.pdf) or [corporate](https://posit.co/wp-content/uploads/2023/04/2023-03-13_TC_Corp_contrib_agreement.pdf) contributor agreement. You can send the signed copy to <jj@rstudio.com>.

Also, please complete and keep the checklist below.

## Description

Please describe your PR here.

````md
```py
print("Hello Quarto!")
```
````

## Checklist

I have (if applicable):

- [ ] filed a [contributor agreement](https://github.com/quarto-dev/quarto-cli/blob/main/CONTRIBUTING.md).
- [ ] referenced the GitHub issue this PR closes
- [ ] updated the appropriate changelog

tarleb committed:
Add `jog` Lua module

tarleb committed:
Allow to use jog instead of walk

Jog can be enable on a per-filter basis by setting the `traverse` field
to `jog`.

tarleb committed:
Rely less on pandoc's type implicit conversions

tarleb committed:
Prevent unintended side-effects of filter functions

tarleb committed:
table-rawhtml: destructively modify the input list

Minor performance improvement.

tarleb committed:
lightbox: Restart imgCount each time the filter is invoked.

This allows to get reproducible output when running the filter multiple
times.

tarleb committed:
Return modified meta object from init filter

If the metadata object is modified in a filter, then it should be
returned, as otherwise the changes might be lost.

tarleb committed:
Fix check for empty captions in parsefiguredivs filters

The `long` field of a caption can never be nil, but it can be an empty
list.

gordonwoodhull commented:
Hi @tarleb, the root cause of the Typst callout failures is that both with jog and without, the following 

https://github.com/quarto-dev/quarto-cli/blob/bf93178346e68f721714ef4c7f792943b4d80f6e/src/resources/filters/customnodes/callout.lua#L46-L51


will initialize `callout.title` to `Inlines({})`

However, when the custom node is read later on to generate Typst, `content.title` is `nil` on main, but 

```
Plain {
  clone: function: 0x600002f273c0
  content: Inlines {}
  show: function: 0x600002f27450
  walk: function: 0x600002f27420
}
```

on jog.

https://github.com/quarto-dev/quarto-cli/blob/bf93178346e68f721714ef4c7f792943b4d80f6e/src/resources/filters/customnodes/callout.lua#L264-L272

It's not the right solution, because we'd like to understand why it is converting (?) differently, but the following diff gets these tests passing:

```diff
diff --git a/src/resources/filters/customnodes/callout.lua b/src/resources/filters/customnodes/callout.lua
index 8a9f4c17d..e9deaa2d2 100644
--- a/src/resources/filters/customnodes/callout.lua
+++ b/src/resources/filters/customnodes/callout.lua
@@ -261,10 +261,14 @@ function _callout_main()
       end
     end
     if callout.attr.identifier == "" then
+      local title = callout.title
+      if not title or #title.content == 0 then
+        title = pandoc.Plain(_quarto.modules.callouts.displayName(callout.type))
+      end
       return _quarto.format.typst.function_call("callout", { 
         { "body", _quarto.format.typst.as_typst_content(callout.content) },
         { "title", _quarto.format.typst.as_typst_content(
-          callout.title or pandoc.Plain(_quarto.modules.callouts.displayName(callout.type))
+          title
         )},
         { "background_color", pandoc.RawInline("typst", background_color) },
         { "icon_color", pandoc.RawInline("typst", icon_color) },
```

tarleb committed:
Use proper metadata types when resolving dependencies

tarleb committed:
Ensure that the internal Meta copy is a true, deep copy

tarleb committed:
Don't add number values to the metadata object

tarleb committed:
Custom nodes: cleanup code for slots/forwarder

tarleb committed:
Fix type of return value in typst filter function

tarleb committed:
Fix type in Caption

tarleb committed:
Rewrite `as_inlines`, `as_blocks` with focus on performance

This also fixes unwanted behavior of `as_blocks`, which would treat a
list of Inline elements as a list of singleton Plain elements, leading
to bad results.

tarleb committed:
Rewrite `as_inlines`, `as_blocks` with focus on performance

This also fixes unwanted behavior of `as_blocks`, which would treat a
list of Inline elements as a list of singleton Plain elements, leading
to bad results.

tarleb committed:
Ensure proper Callout rendering if title is set but empty

tarleb committed:
Ensure proper types when setting custom node slots

tarleb committed:
Add Lua utils function `quarto.utils.is_empty_node`

tarleb committed:
Use the built-in `error` function to trigger a fatal error

tarleb committed:
Provide global function `quarto_assert`

This behaves mostly like the built-in `assert`, but respects the general
Quarto scaffolding.

tarleb committed:
Use `Inlines` metatable for "list of Inline" return values

tarleb committed:
Use global function to check if caption if empty

tarleb committed:
Avoid type confusion with captions during crossref handling

tarleb committed:
Ensure Blocks when normalizing draft posts

tarleb commented:
I've split this into multiple PRs, including #11525, #11526, #11538, and #11540.
